### DISCUSSION TITLE
Avoid cities with BRouter (offline)

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#613] [Q&A] [answered]
New implementation of "Avoid cities" option.

BRouter currently classifies residential areas with population > 50K people.

Old avoid cities|New avoid cities
-|-
<img src="https://github.com/user-attachments/assets/4dcbe683-6cce-4af5-9b6a-0dd49ea1c6d3"/>|<img src="https://github.com/user-attachments/assets/3bdce2f0-9fb0-40fe-873d-08697eca833a"/>

**Comment by DolomitesTom** [comment]
Where do I find this option?

**Reply by devemux86** [reply]
@DolomitesTom 

It is in all route profiles: fastest, curvy, etc. together with the other avoid options.

**Reply by DolomitesTom** [reply]
version didn't update, now ok

**Comment by devemux86** [comment]
BRouter classification of cities > 50K can be seen [here](https://github.com/abrensch/brouter/blob/master/docs/developers/environmental_considerations_and_pseudo_tags.md#town_class).

Now we consider and avoid all town classes the same.

**Comment by umpfmpf** [comment]
![Screenshot_20240830_102548_Cruiser](https://github.com/user-attachments/assets/6797d374-b8a0-4591-afc0-4ecf37a1dc8b)
![Screenshot_20240830_102512_Cruiser](https://github.com/user-attachments/assets/bf29af41-3eb3-4bd2-a83c-46077a8e74fd)
If I plan my route as extra curvy, a route through the wide landscape is calculated. But when I switch on prefer Forests, it suddenly takes me across cities and even though I have set the avoid motorway to 5, it drives 20 km over a motorway.

**Reply by devemux86** [reply]
That's why I still think that the "Prefer" options are not needed and we should remove them!
There were created by BRouter for the cycling and walking profiles.

**Reply by umpfmpf** [reply]
As long as the "prefer" options overlay the avoidance of highways or cities, that is definitely the case. In this case it is unusable (except perhaps for pedestrians or cyclists)

**Reply by devemux86** [reply]
In the cycling and walking profiles the environmental options work completely differently.
Better to remove it again to avoid such confusion and wrong routes.

**Comment by MotoUKRider** [comment]
Here's an interesting comparison - I got the same results in BRouter web.  I turned-off all restrictions and just relied upon the "Avoid cities" in the first image, and turned it off in the second:

![cruiser-avoid-cities](https://github.com/user-attachments/assets/7e629d77-77e3-40b0-b1ba-f5299ffe5a3f)

Interestingly, "Great Dunmow" does not have a population value in OSM, but "Braintree" does (44277). There are a number of possible roads that could have been taken, but the routing really *"went to town"* (if you pardon the pun).  In other places I've tried (and from the example in the top post), the actual result is pretty much what you'd expect.

Knowing the roads, I'm not sure I believe the expected duration of the "scenic" route!  This goes back to earlier posts/woe when there was a considerable difference between an A-B route and a pre-planned GPX route that followed mostly the same path, with the Shaping/Via-Points somehow contributing to the overall journey time.

**Reply by devemux86** [reply]
@MotoUKRider 

Don't forget that BRouter currently classifies residential areas with population > 50K people.

**Reply by MotoUKRider** [reply]
@devemux86 That's the point of the post - There's no population value in OSM for "Great Dunmow".  Wiki says ~10k, and it's a "town" in OSM. So it's neither a "city" nor has > 50k population.  "Bishop's Stortford" is ~41k according to Wiki (nothing on OSM).  Is this a `psuedo_tags` thing?

I appreciate that BRouter is doing this, but it highlights something strange and the unfair thing is to blame the navigation app!

**Reply by devemux86** [reply]
They created "pseudo" tags to host the new classifications.
They use the population data in the OSM, see details [here](https://github.com/abrensch/brouter/blob/master/docs/developers/environmental_considerations_and_pseudo_tags.md).

**Reply by MotoUKRider** [reply]
@devemux86 I was hoping that you'd say: "Yes, this isn't right and BRouter is making things up as it goes along".

I'm concerned that when I use "Avoid cities" (which is meant to apply to places with an OSM population >50k) and BRouter encounters `place=town` and `population=unknown`, it applies the "avoid" rule regardless.  Also, I've just tested a route through the town "Braintree" (pop. 44.2k) and it routes *through* it under normal rules but *around* it when "Avoid cities" is enabled.

So, in short, "Avoid cities" is quite misleading!

**Reply by devemux86** [reply]
@MotoUKRider 

Better test with BRouter website and trekking profile and if there is something strange report the link to their repository.
Otherwise someone needs to study how their "pseudo" tags are created and filled.

**Reply by devemux86** [reply]
> I've just tested a route through the town "Braintree" (pop. 44.2k) and it routes through it under normal rules but around it when "Avoid cities" is enabled.

I cannot reproduce this, a [test](https://brouter.de/brouter-web/#map=12/51.8754/0.5947/standard&lonlats=0.485458,51.804376;0.589846,51.941717) always passes through the city.

Remember that routing is complicated, it also depends where the previous and next waypoints are.

To make a proper test with avoid cities option for a specific residential area,
a small route is needed with 2 waypoints before and after the city, but near the city.
Because if more cities are near the route then they will also affect the result.

**Reply by devemux86** [reply]
"Braintree" has a [node](https://www.openstreetmap.org/node/21503284) (44K) and also a [relation](https://www.openstreetmap.org/relation/52187) (152K).

If we use a [larger route](https://brouter.de/brouter-web/#map=11/51.9689/0.6310/standard&lonlats=0.519618,51.844927;0.706856,52.023533) with waypoints before and after the relation,
then avoid cities option avoids the entire area of the relation.

**Reply by MotoUKRider** [reply]
Interesting.  Your route-link works as expected until I use [the profile you posted here](https://github.com/user-attachments/files/16764042/driving-cruiser.brf.txt) from [the BRouter thread](https://github.com/devemux86/cruiser/discussions/272). I first tried your route on this profile and got the following simply by changing *only* the "consider_town" switch:

`consider_town=unticked` / `consider_town=ticked`
Oddly, we get the opposite outcome here!  The default route takes me around the town via the trunk road (the obvious "fastest" way), but if I tell it to consider the town (eg: avoid cities) then I get routed into it a little bit)...
![brouter-consider-town-1](https://github.com/user-attachments/assets/3dcef091-3490-4673-9f11-7eaec08579c6)

Then I tried my route and got this (and a more sensible outcome):
`consider_town=unticked` / `consider_town=ticked`
![brouter-consider-town-2](https://github.com/user-attachments/assets/de332409-efdc-4172-b447-831849346202)

I see what is happening now, it seems that BRouter is trying to be clever and not sticking just to the `city/town:population=` value but also administrative areas. I can understand what/why it's doing this, because the actual town node doesn't have an "area", so needs to find its boundary from somewhere. My expectations were that it would keep a route away from busy city-centres rather than a regional admin-area as a whole.

**Reply by devemux86** [reply]
This node has population < 50K so it should not be expected to be avoided.

Probably need to check their [SQL](https://github.com/abrensch/brouter/blob/master/docs/developers/environmental_considerations_and_pseudo_tags.md) to see how they classify the residential areas and to what extent.

Maybe @essbee can explain better?

**Reply by devemux86** [reply]
Again, it is best to use only the very detailed street type avoidance options.

Automatic avoidances or preferences may work in one situation and not in another.
See my above [example](https://github.com/devemux86/cruiser/discussions/613#discussion-7035006) where it works nicely.

Have you seen strange routes using the `trekking` or `hiking` profiles?
The environmental options were initially created and tested with these profiles.

You can test them more easily directly on the BRouter website with the original profiles and `consider_town`.

**Comment by devemux86** [comment, marked as answer]
- New implementation of "Avoid cities" option is available in https://github.com/devemux86/cruiser/discussions/641
